### PR TITLE
make it so grapheme can be used as an aa key

### PR DESCRIPTION
**Pull request by schveiguy**
Need an issue report. Just want to make sure it passes CI

### CODE REVIEW

**Comment by PetarKirov** [comment]
@schveiguy the CI is green!

**Review by dkorpel — APPROVED** [review]

**Comment by schveiguy** [comment]
Yeah, I need an issue report. @kinke had a use case (see https://github.com/dlang/phobos/pull/8787#pullrequestreview-1748406488), I'll see if I can test this and make a report tomorrow.

File: std/uni/package.d [hunk 1]
```diff
@@ -7706,6 +7706,12 @@ public:
         return this.tupleof == other.tupleof;
     }
 
+    // Define a default toHash to allow AA usage
+    size_t toHash() const @trusted
+    {
+        return hashOf(slen_, hashOf(small_));
```
Comment (kinke) [inline std/uni/package.d:7712]: AFAICT, this only works for the small-case, not for the heap-buffer case.

Comment (schveiguy) [reply std/uni/package.d:7712]: The heap buffer data doesn't matter. The opEquals is just comparing the pointers.

Not saying this is what we *want* it to be, but that is what it is now.

Comment (kinke) [reply std/uni/package.d:7712]: Yeah I know that the opEquals doesn't work in that case either, and that both implicit opEquals and toHash before v2.106 were most likely working in that case only too (IIRC, unions are by-default compared and hashed as a binary blob of the union's size).

Ideally, we'd be comparing/hashing the underlying bytes slice. But as the `[s]len_` fields seem to be the number of `dchar`s, a proper comparison/hash might have to rely on the `dchar` range returned by `opSlice()`.

Comment (schveiguy) [reply std/uni/package.d:7712]: I don't actually know what the tupleof comparision does for the union...

Comment (schveiguy) [reply std/uni/package.d:7712]: FWIW, I think the proper opEquals and toHash should be based on the opSlice.

Comment (schveiguy) [reply std/uni/package.d:7712]: "it" being the default opEquals or being tupleof == tupleof?

Comment (kinke) [reply std/uni/package.d:7712]: The new `tupleof` comparison. :)

Comment (schveiguy) [reply std/uni/package.d:7712]: wow. I just tested it out. And this is pretty dangerous. For example, if one of the union members (that isn't valid) has an opEquals, then using the tupleof calls this, the default does a memcmp.

In this case of course, it's not a problem. But for sure this is not a good implementation.

Comment (kinke) [reply std/uni/package.d:7712]: Oh, that reminds me of https://issues.dlang.org/show_bug.cgi?id=22171 - the same problem for autogenerated `__xtoHash`, which uses all union members.

**Comment by schveiguy** [comment]
OK, bug report added and ready to close.

**Pull request merged**